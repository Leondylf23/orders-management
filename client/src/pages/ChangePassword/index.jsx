import { showPopup } from '@containers/App/actions';
import { saveNewPassword } from '@pages/Profile/actions';
import { encryptDataAES } from '@utils/allUtils';
import { useState } from 'react';
import { FormattedMessage, useIntl } from 'react-intl';
import { useDispatch } from 'react-redux';
import { Link } from 'react-router-dom';

import ArrowBackIcon from '@mui/icons-material/ArrowBack';

import classes from './style.module.scss';

const ChangePassword = () => {
  const intl = useIntl();
  const dispatch = useDispatch();
  const [userPassword, setUserPassword] = useState({ oldPass: '', newPass: '', confirmPass: '' });

  const saveNewPasswordData = () => {
    if (userPassword?.oldPass === '' || userPassword?.newPass === '' || userPassword?.confirmPass === '') {
      dispatch(
        showPopup(intl.formatMessage({ id: 'profile_title' }), intl.formatMessage({ id: 'profile_password_fill_all' }))
      );
      return;
    }
    if (userPassword?.newPass?.length < 6 || userPassword?.newPass?.length > 20) {
      dispatch(
        showPopup(
          intl.formatMessage({ id: 'profile_title' }),
          intl.formatMessage({ id: 'register_password_validation' })
        )
      );
      return;
    }
    if (userPassword?.newPass !== userPassword?.confirmPass) {
      dispatch(
        showPopup(intl.formatMessage({ id: 'profile_title' }), intl.formatMessage({ id: 'profile_password_same_pass' }))
      );
      return;
    }

    const encryptedData = {
      oldPassword: encryptDataAES(userPassword?.oldPass),
      newPassword: encryptDataAES(userPassword?.newPass),
    };

    dispatch(
      saveNewPassword(
        encryptedData,
        () => {
          setUserPassword({ oldPass: '', newPass: '', confirmPass: '' });
          dispatch(
            showPopup(
              intl.formatMessage({ id: 'profile_title' }),
              intl.formatMessage({ id: 'profile_password_success' })
            )
          );
        },
        () => {
          dispatch(
            showPopup(
              intl.formatMessage({ id: 'profile_title' }),
              intl.formatMessage({ id: 'profile_password_old_pass_not_match' })
            )
          );
        }
      )
    );
  };
  return (
    <div data-testid="changePassword-container">
      <div className={classes.containerTitle}>
        <Link to={-1}>
          <ArrowBackIcon />
        </Link>
        <h3>
          <FormattedMessage id="profile_passwords" />
        </h3>
      </div>
      <label htmlFor="oldPassword" className={classes.label}>
        <FormattedMessage id="profile_password_old" />
      </label>
      <input
        type="password"
        id="oldPassword"
        className={classes.input}
        value={userPassword?.oldPass}
        onChange={(e) => setUserPassword((prevVal) => ({ ...prevVal, oldPass: e.target.value }))}
      />
      <label htmlFor="newPassword" className={classes.label}>
        <FormattedMessage id="profile_password_new" />
      </label>
      <input
        type="password"
        id="newPassword"
        className={classes.input}
        value={userPassword?.newPass}
        onChange={(e) => setUserPassword((prevVal) => ({ ...prevVal, newPass: e.target.value }))}
      />
      <label htmlFor="confirmNewPass" className={classes.label}>
        <FormattedMessage id="profile_password_confirm" />
      </label>
      <input
        type="password"
        id="confirmNewPass"
        className={classes.input}
        value={userPassword?.confirmPass}
        onChange={(e) => setUserPassword((prevVal) => ({ ...prevVal, confirmPass: e.target.value }))}
      />
      <div className={classes.buttonConatainer}>
        <button type="button" className={classes.button} onClick={saveNewPasswordData}>
          <FormattedMessage id="profile_save" />
        </button>
      </div>
    </div>
  );
};

export default ChangePassword;
