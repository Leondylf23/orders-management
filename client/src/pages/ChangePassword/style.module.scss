@import '@styles/base.scss';

.containerTitle {
  margin: 0.5rem 0rem;
  display: flex;
  gap: 20px;
}
.label {
  margin-bottom: 0.25rem;
}
.input {
  @include input;

  width: 100%;
  margin-bottom: 0.5rem;
}
.buttonConatainer {
  @include directionX;
  @include marginY(1rem);

  justify-content: center;

  .button {
    @include button;
  }
}
