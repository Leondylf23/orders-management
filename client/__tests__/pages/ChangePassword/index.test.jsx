import { fireEvent, render } from '@testing-library/react';
import { Provider } from 'react-redux';

import store from '@store';
import Language from '@containers/Language';
import { MemoryRouter } from 'react-router-dom';
import ChangePassword from '@pages/ChangePassword';

jest.mock('react-redux', () => ({
  ...jest.requireActual('react-redux'),
}));

const ParentComponent = (children) => (
  <Provider store={store}>
    <Language>
      <MemoryRouter>{children}</MemoryRouter>
    </Language>
  </Provider>
);

describe('Change Password Page', () => {
  beforeEach(() => {});

  test('Correct Render', () => {
    const changePassword = render(ParentComponent(<ChangePassword />));
    expect(changePassword.getByTestId('changePassword-container')).toBeInTheDocument();
    expect(changePassword).toMatchSnapshot();
  });

  test('Button Clicked', () => {
    const { getByTestId } = render(ParentComponent(<ChangePassword />));
    const buttonSubmit = getByTestId('changePassword-btn');
    expect(buttonSubmit).toBeInTheDocument();
    fireEvent.click(buttonSubmit);
  });
});
