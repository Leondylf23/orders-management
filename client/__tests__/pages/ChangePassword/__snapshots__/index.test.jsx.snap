// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Best Seller Card Correct Render 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        data-testid="changePassword-container"
      >
        <div
          class="containerTitle"
        >
          <a
            href="/"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="ArrowBackIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M20 11H7.83l5.59-5.59L12 4l-8 8 8 8 1.41-1.41L7.83 13H20v-2z"
              />
            </svg>
          </a>
          <h3>
            Kata Sandi
          </h3>
        </div>
        <label
          class="label"
          for="oldPassword"
        >
          Kata Sandi Lama
        </label>
        <input
          class="input"
          id="oldPassword"
          type="password"
          value=""
        />
        <label
          class="label"
          for="newPassword"
        >
          Kata Sandi Baru
        </label>
        <input
          class="input"
          id="newPassword"
          type="password"
          value=""
        />
        <label
          class="label"
          for="confirmNewPass"
        >
          Konfirmasi Kata Sandi
        </label>
        <input
          class="input"
          id="confirmNewPass"
          type="password"
          value=""
        />
        <div
          class="buttonConatainer"
        >
          <button
            class="button"
            type="button"
          >
            Simpan
          </button>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      data-testid="changePassword-container"
    >
      <div
        class="containerTitle"
      >
        <a
          href="/"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="ArrowBackIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M20 11H7.83l5.59-5.59L12 4l-8 8 8 8 1.41-1.41L7.83 13H20v-2z"
            />
          </svg>
        </a>
        <h3>
          Kata Sandi
        </h3>
      </div>
      <label
        class="label"
        for="oldPassword"
      >
        Kata Sandi Lama
      </label>
      <input
        class="input"
        id="oldPassword"
        type="password"
        value=""
      />
      <label
        class="label"
        for="newPassword"
      >
        Kata Sandi Baru
      </label>
      <input
        class="input"
        id="newPassword"
        type="password"
        value=""
      />
      <label
        class="label"
        for="confirmNewPass"
      >
        Konfirmasi Kata Sandi
      </label>
      <input
        class="input"
        id="confirmNewPass"
        type="password"
        value=""
      />
      <div
        class="buttonConatainer"
      >
        <button
          class="button"
          type="button"
        >
          Simpan
        </button>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
